README.txt
MENWIZ: YET ANOTHER CHARACTER LCD MENU WIZARD LIBRARY

This library implement menu creation with simple and compact code.

The pros are:
- (near) asynchronous: that is the library does not take mu control while working, usefull if your sketch need to work while you are changing the settings
- simple to use (few primitives and compact code)
- user defined splash screen (optional): if you need to show something at startup. The time is user defined.
- user defined callback default screen (optional) activated after a user defined elapsed time since last button push. The menu gently leave the screen to the user, untill any button is pushed.  
- you can run a user-defined action (callback routine) from inside the menu
- user defined variables binding: when you set a value inside a menu, the value is changed also in the binded variable available to the sketch code
- the user can overload the built-in  "6 buttons" navigation mode, implementing its own navigation callback (TO BE FULLY TESTED) 

Warnings
- MENWIZ requires New LiquidCrystal (*) and Buttons libraries, enclosed with the package. The New LiquidCrystal allows to use 4 wires, 8 wires, I2C, serial and other LCDs. The use of New LiquidCrystal is beyond this introduction
- Menu labels require memory. Be carefull not to run out of  memory (the library provides a simple memory-check function: use it!)
- the examples are done for 20x4 I2C LCD. other interfaces and screen dimensions can be used accordingly with New LiquidCrystal interface (changing mode/interface does not affect library code, but only sketch code) 

The MENWIZ library is avalable at:
gethub.com/brunialti/MENWIZ.git

Any feed back is welcome. Should you be interested I can mantain and expand.
Keep in mind that I want to keep it simple.

Next steps are:
- new data types
- code inspection and cleaning

The version number convention is as following:
<version>.<subversion>.<build>
  <version>: 0=not officially released, you have to check the latest available sources, as the lib could not work properly
  <subversion>: new functions, modified function interfaces
  <build>: bug fixing, minor internal changes


The TEST_LIB zip file contains the library used to test the MENWIZ lib.
I used what I felt to be "standard" libraries. Should you have some problem using yours actual libs, try using them

(*) New LiquidCrystal is a replacement of the standard LiquidCrystal and LiquidCrystal_I2C libraries. it is more efficient and very well supported
That's the nice thing about the "New LiquidCrystal" library, it already supports a wide range of LCD controllers:
- I2C base on the PCF8574 or compatible
- 4bit and 8bit parallel
- ShiftRegister: 3wire, 2wire, 1wire (beta)
please have a look at http://arduino.cc/forum/index.php/topic,77120.0.html
You can load the latest version at  https://bitbucket.org/fmalpartida/new-liquidcrystal/wiki/Home



CHANGES.txt
*********************************
Ver 0.4.1 CHANGES ***************
*********************************
 
*** CHANGES TO EXISTING FUNCTIONS

addVar(int,float *,float,float,float)
now MENWIZ supports variables of floating point type (MW_AUTO_FLOAT). The variables are displayed with a nember of decimal digits set by MW_FLOAT_DEC global variable (default=1). The syntax is the same as integer type (MW_AUTO_INTEGER).
Example:
float gp;
menu.addVar(MW_AUTO_FLOAT,&gp,11.00,100.00,0.5);
the above call create a variable of type float, binded to sketch variable gp, ranging between 11,0 and 100,0, with increment of 0,5

addVar(int,byte *,byte,byte,byte)
now MENWIZ supports now also variables of byte type (MW_AUTO_BYTE). The syntax is the same as integer type (MW_AUTO_INTEGER).
Example:
byte gp;
menu.addVar(MW_AUTO_BYTE,&gp,0,255,1);
the above call create a variable of type byte, binded to sketch variable gp, ranging between 1,0 and 255, with increment of 1

*** NEW FUNCTIONS ***************

*** BUG FIXED *******************

*** INTERNAL CHANGES ************

added the global variable MW_FLOAT_DEC

*********************************
Ver 0.3.0 CHANGES *************** 
*********************************

*** CHANGES TO EXISTING FUNCTIONS

addSplash
the string passed to the function use \n (0x0A) character as line delimiter instead of '#'
  

*** NEW FUNCTIONS ***************

addUsrNav (WARNING: NOT FULLY TESTED YET)
user defined navigation device routine (callback). The user can use any device other than buttons to overwrite the internal routine. The callback *must* return an int code for any pushed "button" (MW_BTU=UP,MW_BTD=DOWN,MW_BTL=LEFT,MW_BTR=RIGHT,MW_BTE=ESCAPE,MW_BTC=CONFIRM,MW_BTNULL=NO BUTTON). 
The callback is invocated on each call to the method draw. The used device(s) must be declared and initialized inside the sketch by the user. The callback is in charge of  device debouncing (if any).

drawUsrScreen
added public method to class menwiz. It quick draw LCD screen with the contents of the argument string. Each line to be shown in the LCD is terminated by char 0x0A ('\n') inside the argument string. This method provide the user with the quick way to write an entire LCD screen (the lib will manage space padding, cursor position and string length checking) 
Example: 
menu.drawUsrScreen("Test user screen\nline1\nline2\n\n"); 

getErrorMessage
added a boolean argument: if true, the function write a full error message to the default serial terminal, otherwise return error code only 

*** BUG FIXED *******************


*** INTERNAL CHANGES ************

- abstracted the methods for button checking, in order to add aaUsrNav functions
- reinforced error checkin drawUsrScreen (string truncation)
- renaming internal variables



MENWIZ.h
// ---------------------------------------------------------------------------
// Created by Roberto Brunialti on 20/04/12.
// Copyright 2012 - Under creative commons license 3.0:
//        Attribution-ShareAlike CC BY-SA
//
// This software is furnished "as is", without technical support, and with no 
// warranty, express or implied, as to its usefulness for any purpose.
//
// Thread Safe: No
// Extendable: Yes
// 
// @file MENWIZ.h
// This file implements a basic menu management library in the Arduino SDK
// 
// @brief 
// This is a menu management library. The library allows user to create
// an intire menu tree with relatively few lines of code.
// The library allows the users to define callbacks able to overload internal
// functions (i.e. navigation device management) or to be executed inside a menu.
// It is possible to define also splash screens and a default user screen to be
// activated after a time interval since the last interactio,
//
// @author R. Brunialti - roberto_brunialti@tiscali.it
// ---------------------------------------------------------------------------

#ifndef MENWIZ_h
#define MENWIZ_h

#include <Wire.h>
#include <LCD.h>
#include <buttons.h>

#if (ARDUINO >= 100)
  #include <Arduino.h> 
#else
  #include <WProgram.h>
#endif 

// DIMENSIONS (DIMENSIONAL LIMITS OF STATICALLY ALLOCATED STRUCTURES)
// ---------------------------------------------------------------------------
#define MAX_MENU       15   //maximum number of nodes (absolute supported max number of addMenu calls)
#define MAX_OPTXMENU   5    //maximum number of options/submenus for each node (max number of addItem call for each menu item) 
#define MAX_BUFFER     84   //dimension=columns x rows + rows. CHANGE IT IF LCD IS BIGGER THAN 4X20 CHARS

// NODE TYPES 
// ---------------------------------------------------------------------------
#define MW_ROOT        10  //root menu
#define MW_SUBMENU     11  //submenu
#define MW_VAR         12  //terminal node

// CUR MODE
// ---------------------------------------------------------------------------
#define MW_MODE_SPLASH     20  
#define MW_MODE_USRSCREEN  21  
#define MW_MODE_MENU       22  

// BUTTON CODES
// ---------------------------------------------------------------------------
#define MW_BTNULL      30   //NOBUTTON
#define MW_BTU         31   //UP
#define MW_BTD         32   //DOWN
#define MW_BTL         33   //RIGTH
#define MW_BTR         34   //LEFT
#define MW_BTE         35   //ESCAPE
#define MW_BTC         36   //CONFIRM

// VALUE TYPES
// ---------------------------------------------------------------------------
#define MW_LIST        41  //OPTION LIST
#define MW_BOOLEAN     42  //ON/OFF TOGGLE
#define MW_AUTO_INT    43  //INTEGER VALUE WITH INCREMENT STEP
#define MW_AUTO_FLOAT  44  //FLOATING POINT VALUE WITH INCREMENT STEP
#define MW_AUTO_BYTE   45  //byte VALUE WITH INCREMENT STEP
#define MW_TEXT        46  //not implemented yet
#define MW_ACTION      47  //FIRE AN ACTION WHEN CONFIRM BUTTON IS PUSHED
#define MW_EDIT_INT    48  //not implemented yet
#define MW_EDIT_FLOAT  49  //not implemented yet

// DEREFERENCING OPERATORS
// ---------------------------------------------------------------------------
#define VBOOL(a)    *(boolean*)a
#define VINT(a)     *(int*)a
#define VBYTE(a)    *(byte*)a
#define VFLOAT(a)   *(float*)a
#define VFUNC(a)    (* a)

// FLAGS
// ---------------------------------------------------------------------------
#define FL_SPLASH 		1   		
#define FL_SPLASH_DRAW		2 	
#define FL_USRSCREEN_DRAW 	3  	

// OTHERS
// ---------------------------------------------------------------------------
#define MW_EOL_CHAR    0x0A
#define MW_TYPE byte

typedef struct{
  boolean fl;
  void (*fv)();
  int  (*fi)();
}_cback;

typedef struct{
  int last_button;
  unsigned long tm_push;
  Button    BTU;   
  Button    BTD;   
  Button    BTL;   
  Button    BTR;   
  Button    BTE;   
  Button    BTC;   
}_nav;

class _var{
public:
           _var();

  MW_TYPE  type;
  void*    val;
  void*    old;
  void     (*action)();
  void*    lower;
  void*    upper;
  void*    incr;   
};

class _option{
public:
           _option();

  MW_TYPE  type;
  char*    label;
  byte     sbm;  //submemu id if type=SUBMENU
private:
protected:
};

class _menu{
public:
           _menu();
  void     addVar(int, int*);
  void     addVar(int, int*, int, int, int);
  void     addVar(int, float*, float, float, float);
  void     addVar(int, byte *,byte ,byte ,byte);
  void     addVar(int, boolean *);
  void     addVar(int, void (*f)());
  _option* addItem(int, char*);

  MW_TYPE  type;
  _var     var;
  char*    label;
  byte     cod;
  byte     parent;
  byte     cur_item;
  byte     idx_o;    //option index
  void*    o[MAX_OPTXMENU];
private:
protected:
};

class menwiz{
public:
           menwiz();
  void     begin(void *,int, int);
  void     addSplash(char *,int);
  void     addUsrScreen(void (*f)(), unsigned long);
  void     addUsrNav(int (*f)());
  _menu*   addMenu(int, _menu *, char *);
  void     draw();
  void     navButtons(int,int,int,int,int,int);
  void     drawUsrScreen(char *);       //draw user screen(s)
  int      getErrorMessage(boolean); 	//if arg=true, err message is printed to the default Serial terminal, otherwise the function returns error code only
  int      freeRam();
  char*    getVer();
  byte     flags;
  LCD *    lcd;
  char     sbuf[MAX_BUFFER];            //lcd screen buffer (+ 1 for each line) 
  unsigned long tm_start;       	//start time (set when begin method is invocated)
  unsigned long tm_splash;      	//splash screen duration  
  unsigned long tm_usrScreen;   	//lap time before usrscreen  
  _cback   usrScreen;	        	//callback
  _cback   usrNav;    
  byte     idx_m;
  _menu    m[MAX_MENU];
  _menu*   cur_menu;
  _menu*   root;
  _nav     btx;
private:
  byte     row;
  byte     col;
  byte     cur_mode;
  int      scanNavButtons();
  int      actNavButtons(int);
  void     drawMenu(_menu *);
  void     drawVar(_menu *);
  void     actBTU();
  void     actBTD();
  void     actBTL();
  void     actBTR();
  void     actBTE();
  void     actBTC();
protected:
};

#endif




MENWIZ.cpp
// ---------------------------------------------------------------------------
// Created by Roberto Brunialti on 20/04/12.
// Copyright 2012 - Under creative commons license 3.0:
//        Attribution-ShareAlike CC BY-SA
//
// This software is furnished "as is", without technical support, and with no 
// warranty, express or implied, as to its usefulness for any purpose.
//
// Thread Safe: No
// Extendable: Yes
// 
// @file MENWIZ.cpp
// This file implements a basic menu management library in the Arduino SDK
// 
// @brief 
// This is a menu management library. The library allows user to create
// an intire menu tree with relatively few lines of code.
// The library allows the users to define callbacks able to overload internal
// functions (i.e. navigation device management) or to be executed inside a menu.
// It is possible to define also splash screens and a default user screen to be
// activated after a time interval since the last interaction
//
// @author R. Brunialti - roberto_brunialti@tiscali.it
// ---------------------------------------------------------------------------
#include "MENWIZ.h"

#define SCREATE(p,s) p=(char *)malloc(strlen(s)+1); strcpy(p,s)
#define SFORM(b,s,l) memset(b,32,l); memcpy(b,s,strlen(s)); b[l]=NULL; lcd->print(b)
#define ERROR(a)     MW_error=a
#define FILLBLANK(b,l) memset(b,32,l);b[l]=NULL; lcd->print(b)

// GLOBAL VARIABLES
// ---------------------------------------------------------------------------
int MW_FLOAT_DEC=1;
char buf[81];
static const char MW_ver[]={"0.4.2"};
static const char MW_FMT_VARINT[]={"%d [%d] %d"};
static const char MW_STR_CONFIRM[]={"[C] to run."};
static const uint8_t c0[8]={B00000, B00000, B00001, B00010, B10100, B01000, B00000, B00000}; 
byte MW_error;

menwiz::menwiz(){
  bitWrite(flags,FL_SPLASH,0);
  bitWrite(flags,FL_SPLASH_DRAW,0);
  bitWrite(flags,FL_USRSCREEN_DRAW,0);
  usrScreen.fl=false;
  usrNav.fl=false;
  btx.last_button=MW_BTU;
  idx_m=0;
  row=0;
  col=0;
  tm_start=0;  
  root=NULL;
}

_menu::_menu(){
  idx_o=0;
  cur_item=0;
  label=NULL;
  parent=NULL;
}

_option::_option(){
  label=NULL;
}

_var::_var(){
  val=NULL;
  old=NULL;
  lower=NULL;
  upper=NULL;
  incr=NULL;   
  type=0;
}

_menu * menwiz::addMenu(int t,_menu * p, char *lab){
   _option *op;

  ERROR(0);   
   //INSTANTIATE NEW MENU VARIABLES
  if (idx_m<MAX_MENU){   
    m[idx_m].type=(MW_TYPE)t;         // ROOT| SUBMENU| VAR
    SCREATE(m[idx_m].label,(char *)lab);      // OPT:STORE STRING BY REFERENCE, NOT BY VALUE?
    m[idx_m].cod=idx_m;      // unique menu id
    if (t==MW_ROOT){
      //IF ROOT, PARENT=ITSELF, SET ROOT POINTER, SET ROOT AS START MENU 
      m[idx_m].parent=idx_m;
      root=&m[idx_m];
      cur_menu=&m[idx_m];
      }
    else{
      //IF NOT ROOT, ADD MENU TO THE PARENTS OPLIST 
      m[idx_m].parent=p->cod;
      op=p->addItem(t,(char *)lab);
      op->sbm=idx_m;
      }
    idx_m++;
    return &m[idx_m-1];
    }
  else{ERROR(100);}
// ERROR    
   }

_option *_menu::addItem(int t,char *lab){
  static _option *op;

  ERROR(0);
  if (idx_o<MAX_OPTXMENU){
    op=(_option *) malloc(sizeof(_option));
    op->type=(MW_TYPE)t;
    SCREATE(op->label,(char *)lab);
    o[idx_o]=(_option*)op;
    idx_o++;
    }
  else{ERROR(105);}
// ERROR    
   } 

void _menu::addVar(int t, int* v){

  ERROR(0);
  if(type==MW_ROOT){     //patch to be verified
    type=(MW_TYPE)MW_VAR;    //patch to be verified
    }	
  if (t!=MW_LIST)
    ERROR(120);
  else if(type==MW_VAR){
    var.type=(MW_TYPE)t;
    var.val=v;
    cur_item=VBYTE(v);
    }
  else{ERROR(110);}
// ERROR    
  }
  
void _menu::addVar(int t, int* v, int low, int up, int incr){

  ERROR(0);
  if (t!=MW_AUTO_INT)
    ERROR(120);
  else if(type==MW_VAR){
    var.type=(MW_TYPE)t;
    var.val=v;
    var.lower=malloc(sizeof(int)); if(var.lower!=NULL) VINT(var.lower)=low; else {ERROR(900); return;}
    var.upper=malloc(sizeof(int)); if(var.upper!=NULL) VINT(var.upper)=up; else {ERROR(900); return;}
    var.incr=malloc(sizeof(int));  if(var.incr!=NULL) VINT(var.incr)=incr; else {ERROR(900); return;} 
    var.old=malloc(sizeof(int));   if(var.old!=NULL) VINT(var.old)=VINT(var.val); else {ERROR(900); return;} 
    }
  else{ERROR(110);}
// ERROR    
  }

void _menu::addVar(int t, float* v, float low, float up, float incr){

  ERROR(0);
  if (t!=MW_AUTO_FLOAT)
    ERROR(120);
  else if(type==MW_VAR){
    var.type=(MW_TYPE)t;
    var.val=v;
    var.lower=malloc(sizeof(float)); if(var.lower!=NULL) VFLOAT(var.lower)=low; else {ERROR(900); return;}
    var.upper=malloc(sizeof(float)); if(var.upper!=NULL) VFLOAT(var.upper)=up; else {ERROR(900); return;}
    var.incr=malloc(sizeof(float));  if(var.incr!=NULL) VFLOAT(var.incr)=incr; else {ERROR(900); return;} 
    var.old=malloc(sizeof(float));   if(var.old!=NULL) VFLOAT(var.old)=VFLOAT(var.val); else {ERROR(900); return;} 
    }
  else{ERROR(110);}
// ERROR    
  }

void _menu::addVar(int t, byte* v, byte low, byte up, byte incr){

  ERROR(0);
  if (t!=MW_AUTO_BYTE)
    ERROR(120);
  else if(type==MW_VAR){
    var.type=(MW_TYPE)t;
    var.val=v;
    var.lower=malloc(sizeof(byte)); if(var.lower!=NULL) VBYTE(var.lower)=low; else {ERROR(900); return;}
    var.upper=malloc(sizeof(byte)); if(var.upper!=NULL) VBYTE(var.upper)=up; else {ERROR(900); return;}
    var.incr=malloc(sizeof(byte));  if(var.incr!=NULL) VBYTE(var.incr)=incr; else {ERROR(900); return;} 
    var.old=malloc(sizeof(byte));   if(var.old!=NULL) VBYTE(var.old)=VBYTE(var.val); else {ERROR(900); return;} 
    }
  else{ERROR(110);}
// ERROR    
  }

void _menu::addVar(int t, boolean* v){

  ERROR(0);
  if (t!=MW_BOOLEAN)
    ERROR(120);
  else if(type==MW_VAR){
    var.type=(MW_TYPE)t;
    var.val=v;
    var.old=malloc(sizeof(boolean));  if(var.old!=NULL) VBOOL(var.old)=VBOOL(var.val); else {ERROR(900); return;} 
    }
  else{ERROR(110);}
// ERROR    
  }

void  _menu::addVar(int t,void (*f)()){

  ERROR(0);
  if (t!=MW_ACTION)
    ERROR(120);
  else if(type==MW_VAR){
    var.type=(MW_TYPE)t;
    var.action=f;
    }
  else{ERROR(110);}
// ERROR    
  }

void menwiz::addUsrNav(int (*f)()){
  usrNav.fl=true;
  usrNav.fi=f;
  }

void menwiz::begin(void *l,int c, int r){

  ERROR(0);
  tm_start=millis();
  row=r;
  col=c;
  flags=0;
  lcd=(LCD*)l; 
  lcd->begin(c,r);  // Size of the LCD
  lcd->setBacklight(HIGH);
  lcd->noCursor();
  lcd->createChar(0,(uint8_t*)c0);
//  sbuf=(char*)malloc(r*c+r); if(sbuf==NULL) ERROR(900);
//  buf=(char*)malloc(c+1); if(buf==NULL) ERROR(900);  
//  bitWrite(flags,FL_USRSCREEN_DRAW,0);
//  bitWrite(flags,FL_SPLASH_DRAW,0);
  }

char* menwiz::getVer(){

  ERROR(0);
  return (char*) MW_ver; 
  }

void menwiz::draw(){
  int ret;

  ERROR(0);
  // get nav choice
  ret= usrNav.fl?ret=usrNav.fi():scanNavButtons();    	//internal method or user defined callback?
  // if usrscreen is active, skip last button and switch to MENU mode
  if((cur_mode==MW_MODE_USRSCREEN)&&(ret!=MW_BTNULL)){
    cur_mode=MW_MODE_MENU;
    btx.last_button=MW_BTNULL;
    btx.tm_push=millis();
    } 
  // else run the action associated to selected button
  else
    ret=actNavButtons(ret);
  
  int long lap1=(millis()-tm_start);
  int long lap2=(millis()-btx.tm_push);

  // if defined splashscreen & not yet drawn & time window is ok, draw it  
  if((bitRead(flags,FL_SPLASH)==true) && (lap1<tm_splash)){
    cur_mode=MW_MODE_SPLASH;
    //draw only once
    if(bitRead(flags,FL_SPLASH_DRAW)==false){
      drawUsrScreen(sbuf);
      bitWrite(flags,FL_SPLASH_DRAW,1);
      }
    }
  // if defined usrscreen & time since last button push > user defined time, draw it  
  else if((usrScreen.fl) && (lap2>tm_usrScreen)){
    cur_mode=MW_MODE_USRSCREEN;
    bitWrite(flags,FL_USRSCREEN_DRAW,0);
    usrScreen.fv();
    }
  else{
  // if a button was pushed since last call, draw menu  
    cur_mode=MW_MODE_MENU;
    if((btx.last_button!=MW_BTNULL) || (!bitRead(flags,bitRead(flags,FL_USRSCREEN_DRAW))))
      bitWrite(flags,FL_USRSCREEN_DRAW,1);
      drawMenu(cur_menu);
    }
  }

void menwiz::drawUsrScreen(char *scr){
  int i,j,k;

  ERROR(0);
  for (int i=0,j=0,k=0;(k<row)&&(i<(strlen(scr)+1));i++){
    if((scr[i]==MW_EOL_CHAR)||(scr[i]==0)){
      memset(&buf[min(j,col)],' ',col-j);
      buf[col]=0;
      lcd->setCursor(0,k);
      lcd->print(buf);
      j=0;
      k++;
      }
    else{
      buf[j]=scr[i];
      j++;
      }
    if(k==row)
      return;
    }
  }
  
void menwiz::drawMenu(_menu *mc){
  int rstart,rstop,i,j;
  _option *op;

  ERROR(0);
  lcd->setCursor(0,0);
  SFORM(buf,(char*)mc->label,(int) col);

  if (mc->type==MW_VAR){
    drawVar(mc);
    }
  else{
    rstart=max(0,mc->cur_item-(row-2));
    rstop=min((mc->idx_o),(rstart+row));
    for (i=1,j=rstart;i<row;i++,j++){// for all remaining lcd rows
      if(j<rstop){
        op=(_option*)mc->o[j];
        lcd->setCursor(0,i);
        lcd->write((j==mc->cur_item)?126:165);
        SFORM(buf,op->label,(int) col-1);
        }
      else{
        lcd->setCursor(0,i);
        FILLBLANK(buf,col);
        }
      }
    }    
  }

void menwiz::drawVar(_menu *mc){
  int rstart,rstop,i,j;
  _option *op;
  
  ERROR(0);
  switch ((mc->var.type)){
    case MW_LIST:
      rstart=max(0,mc->cur_item-(row-2));
      rstop=min((mc->idx_o),rstart+(row));
      for (i=1,j=rstart;i<row;i++,j++){// for all remaining lcd rows
        if(j<rstop){
          op=(_option*)mc->o[j];
          lcd->setCursor(0,i);
          lcd->write((j==mc->cur_item)?0:165);
          SFORM(buf,op->label,(int) col-1);
          }
        else{
          lcd->setCursor(0,i);
          FILLBLANK(buf,col);
          }
        }
      break;  
    case MW_AUTO_INT:
    case MW_AUTO_BYTE:
      for(i=2;i<row;i++){
        lcd->setCursor(0,i);
        FILLBLANK(buf,col);
        }
      lcd->setCursor(0,1);
      if(mc->var.type==MW_AUTO_INT)
      	sprintf(sbuf,MW_FMT_VARINT,VINT(mc->var.lower),VINT(mc->var.old),VINT(mc->var.upper));
      else
      	sprintf(sbuf,MW_FMT_VARINT,VBYTE(mc->var.lower),VBYTE(mc->var.old),VBYTE(mc->var.upper));
      SFORM(buf,sbuf,(int) col);
      break;      
    case MW_AUTO_FLOAT:
      for(i=2;i<row;i++){
        lcd->setCursor(0,i);
        FILLBLANK(buf,col);
        }
      lcd->setCursor(0,1);
      lcd->print(dtostrf(VFLOAT(mc->var.lower),0,MW_FLOAT_DEC,buf));
      lcd->print(" [");
      lcd->print(dtostrf(VFLOAT(mc->var.old),0,MW_FLOAT_DEC,buf));
      lcd->print("] ");
      lcd->print(dtostrf(VFLOAT(mc->var.upper),0,MW_FLOAT_DEC,buf));
      break;      
    case MW_BOOLEAN:
      for(i=2;i<row;i++){
        lcd->setCursor(0,i);
        FILLBLANK(buf,col);
        }
      lcd->setCursor(0,1);
      SFORM(buf,VBOOL(mc->var.old)?"ON":"OFF",(int) col);
      break;      
    case MW_ACTION:
      for(i=2;i<row;i++){
        lcd->setCursor(0,i);
        FILLBLANK(buf,col);
        }
      lcd->setCursor(0,1);
      SFORM(buf,MW_STR_CONFIRM,(int) col);
      break;      
    default:{}
    }
  }

void menwiz::addSplash(char *s, int lap){

  ERROR(0);
  strcpy(sbuf,s);
  tm_splash=lap;
  bitWrite(flags,FL_SPLASH,1);
  bitWrite(flags,FL_SPLASH_DRAW,0);
  }

void menwiz::addUsrScreen(void f(), unsigned long t){

  ERROR(0);
  usrScreen.fv=f;
  usrScreen.fl=true;
  tm_usrScreen=t;
  }

void menwiz::navButtons(int btu,int btd,int btl,int btr,int bte,int btc){
  
  ERROR(0);
  if(btu!=0){btx.BTU.assign(btu);  btx.BTU.setMode(OneShot);  btx.BTU.turnOnPullUp();} 
  if(btd!=0){btx.BTD.assign(btd);  btx.BTD.setMode(OneShot);  btx.BTD.turnOnPullUp();} 
  if(btl!=0){btx.BTL.assign(btl);  btx.BTL.setMode(OneShot);  btx.BTL.turnOnPullUp();} 
  if(btr!=0){btx.BTR.assign(btr);  btx.BTR.setMode(OneShot);  btx.BTR.turnOnPullUp();} 
  if(bte!=0){btx.BTE.assign(bte);  btx.BTE.setMode(OneShot);  btx.BTE.turnOnPullUp();} 
  if(btc!=0){btx.BTC.assign(btc);  btx.BTC.setMode(OneShot);  btx.BTC.turnOnPullUp();} 

  // bouncing disarm 
  btx.BTU.check();
  btx.BTD.check(); 
  btx.BTL.check(); 
  btx.BTR.check(); 
  btx.BTE.check(); 
  btx.BTC.check(); 
  }

int menwiz::scanNavButtons(){ 
  if(btx.BTU.check()==ON){
    return MW_BTU;}
  else if (btx.BTD.check()==ON){
    return MW_BTD;}
  else if (btx.BTL.check()==ON){
    return MW_BTL;}
  else if (btx.BTR.check()==ON){
    return MW_BTR;}
  else if (btx.BTE.check()==ON){
    return MW_BTE;}
  else if (btx.BTC.check()==ON){
    return MW_BTC;}
  else
    return MW_BTNULL;
  }

int menwiz::actNavButtons(int button){  
  ERROR(0);
  if (button==MW_BTNULL) 
    return(button);
  else{
    btx.last_button=button;
    btx.tm_push=millis();
    switch(button){
      case MW_BTU: actBTU();  break; 
      case MW_BTD: actBTD();  break; 
      case MW_BTL: actBTL();  break; 
      case MW_BTR: actBTR();  break; 
      case MW_BTE: actBTE();  break; 
      case MW_BTC: actBTC();  break; 
      }
    }
  }

void menwiz::actBTU(){ 
  if((cur_menu->type!=MW_VAR)||(cur_menu->var.type==MW_LIST))    
    cur_menu->cur_item=(cur_menu->cur_item-1)<0?(cur_menu->idx_o-1):cur_menu->cur_item-1;
  }

void menwiz::actBTD(){ 
  if((cur_menu->type!=MW_VAR)||(cur_menu->var.type==MW_LIST))    
    cur_menu->cur_item=(cur_menu->cur_item+1)%(cur_menu->idx_o);
  }

void menwiz::actBTL(){ 
  if(cur_menu->var.type==MW_AUTO_BYTE){
    VBYTE(cur_menu->var.old)=max((VBYTE(cur_menu->var.old)-VBYTE(cur_menu->var.incr)),VBYTE(cur_menu->var.lower));
    }
  else if(cur_menu->var.type==MW_AUTO_INT){
    VINT(cur_menu->var.old)=max((VINT(cur_menu->var.old)-VINT(cur_menu->var.incr)),VINT(cur_menu->var.lower));
    }    
  else if(cur_menu->var.type==MW_AUTO_FLOAT){
    VFLOAT(cur_menu->var.old)=max((VFLOAT(cur_menu->var.old)-VFLOAT(cur_menu->var.incr)),VFLOAT(cur_menu->var.lower));
    }    
  else if(cur_menu->var.type==MW_BOOLEAN){
    VBOOL(cur_menu->var.old)=!VBOOL(cur_menu->var.old);
    }    
  }

void menwiz::actBTR(){ 
  if(cur_menu->var.type==MW_AUTO_INT){
    VINT(cur_menu->var.old)=min((VINT(cur_menu->var.old)+VINT(cur_menu->var.incr)),VINT(cur_menu->var.upper));
    }    
  else if(cur_menu->var.type==MW_AUTO_BYTE){
    VBYTE(cur_menu->var.old)=min((VBYTE(cur_menu->var.old)+VBYTE(cur_menu->var.incr)),VBYTE(cur_menu->var.upper));
    }    
  else if(cur_menu->var.type==MW_AUTO_FLOAT){
    VFLOAT(cur_menu->var.old)=min((VFLOAT(cur_menu->var.old)+VFLOAT(cur_menu->var.incr)),VFLOAT(cur_menu->var.upper));
    }    
  else if(cur_menu->var.type==MW_BOOLEAN){
    VBOOL(cur_menu->var.old)=!VBOOL(cur_menu->var.old);
    }    
  }

void menwiz::actBTE(){ 
  if((cur_menu->type==MW_VAR)&&(cur_menu->var.type==MW_AUTO_INT)){        
    VINT(cur_menu->var.old)=VINT(cur_menu->var.val);
    }
  else if((cur_menu->type==MW_VAR)&&(cur_menu->var.type==MW_AUTO_FLOAT)){        
    VFLOAT(cur_menu->var.old)=VFLOAT(cur_menu->var.val);
    }
  else if((cur_menu->type==MW_VAR)&&(cur_menu->var.type==MW_AUTO_BYTE)){        
    VBYTE(cur_menu->var.old)=VBYTE(cur_menu->var.val);
    }
  if((cur_menu->type==MW_VAR)&&(cur_menu->var.type==MW_BOOLEAN)){        
    VBOOL(cur_menu->var.old)=VBOOL(cur_menu->var.val);
    }
  cur_menu=&m[cur_menu->parent];
  }

void menwiz::actBTC(){ 
  _option* oc;
  oc=(_option*)cur_menu->o[cur_menu->cur_item];

  if((cur_menu->type==MW_SUBMENU)||(cur_menu->type==MW_ROOT)){
    cur_menu=&m[oc->sbm];
    if((cur_menu->type==MW_VAR)&&(cur_menu->var.type==MW_LIST))        
      cur_menu->cur_item=VINT(cur_menu->var.val);
      }
  else if((cur_menu->type==MW_VAR)&&(cur_menu->var.type==MW_LIST)){
    VINT(cur_menu->var.val)=(int)cur_menu->cur_item;
    cur_menu=&m[cur_menu->parent];
    }
  else if((cur_menu->type==MW_VAR)&&(cur_menu->var.type==MW_AUTO_FLOAT)){        
    VFLOAT(cur_menu->var.val)=VFLOAT(cur_menu->var.old);
    cur_menu=&m[cur_menu->parent];
    }
  else if((cur_menu->type==MW_VAR)&&(cur_menu->var.type==MW_AUTO_INT)){        
    VINT(cur_menu->var.val)=VINT(cur_menu->var.old);
    cur_menu=&m[cur_menu->parent];
    }
  else if((cur_menu->type==MW_VAR)&&(cur_menu->var.type==MW_AUTO_BYTE)){        
    VBYTE(cur_menu->var.val)=VBYTE(cur_menu->var.old);
    cur_menu=&m[cur_menu->parent];
    }
  else if((cur_menu->type==MW_VAR)&&(cur_menu->var.type==MW_BOOLEAN)){        
    VBOOL(cur_menu->var.val)=VBOOL(cur_menu->var.old);
    cur_menu=&m[cur_menu->parent];
    }
  else if((cur_menu->type==MW_VAR)&&(cur_menu->var.type==MW_ACTION)){        
    cur_menu->var.action();	      
    cur_menu=&m[cur_menu->parent];
    }
  }

int menwiz::getErrorMessage(boolean fl){
  if (fl){
    switch(MW_error)
      {
      case 0:   break; 
      case 100: Serial.println(F("ERR.100-Too many _menu items"));break; 
      case 105: Serial.println(F("ERR.105-Too many _option items"));break; 
      case 110: Serial.println(F("ERR.110-MW_VAR menu type required"));break; 
      case 120: Serial.println(F("ERR.120-Bad 1st arg"));break; 
      case 900: Serial.println(F("ERR.900-Out of memory"));break; 
      default:  Serial.println(F("ERR.000-Unknown err"));break; 
      }
    }
  return MW_error;
  }

int menwiz::freeRam () {
  extern int __heap_start, *__brkval; 
  int v; 
  return (int) &v - (__brkval == 0 ? (int) &__heap_start : (int) __brkval); 
  }

